composer/checkpoint/save.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Useful functions for saving state dicts to disk."""
+
+import logging
+import os
+import textwrap
+import warnings
+from pathlib import Path
+from typing import Any, Dict, Optional, Union
+
+import torch.distributed.checkpoint as DCP
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+from torch.distributed._tensor import DTensor
+
+from composer.utils import dist
+from composer.utils.checkpoint import _TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME, _write_checkpoint_file
+
+log = logging.getLogger(__name__)
+
+
+def save_state_dict_to_disk(
+    state_dict: Dict[str, Any],
+    destination_file_path: str,
+    overwrite: bool = False,
+    save_format: str = 'pt',  # or hf, safetensor
+) -> Optional[str]:
+    """Saves a state dict to local disk.
+
+    Args:
+        state_dict (Dict[str,Any]): The state dict to save.
+        destination_dir (str): The directory to save the state dict to.
+        filename (str): The name of the file to save the state dict to.
+        overwrite (bool): If True, the file will be overwritten if it exists.
+        save_format (str): The format to save the state dict in. One of 'pt', 'hf', or 'safetensor'.
+        async_save (bool): If True, the save will be done asynchronously and the function will return with the path of where it was going to be saved
+
+    Returns:
+        str: The full path to the saved state dict if sharded is false and rank 0 or if sharded is true, otherwise None.
+    """
+    if state_dict == {}:
+        return None
```

Comment:
why? can you not save an empty state_dict?

Reply:
why would we save an empty state dict?

Reply:
the use case for this is if you generate a full state dict and pass it to the save function then all non-zero-ranks will have an empty state dict, so either:
* we save an empty file for every rank. I can't think of a use case where you'd want that
* we require the user to add a gating by rank if statement (the idea is that the user doesn't have to make any dist.* calls nor do they even have to know about dist.*)